src/psm/src/ir_network.h

```diff
@@ -0,0 +1,213 @@
+  void dumpNodes(const std::map<Node*, std::size_t>& node_map,
+                 const std::string& name = "nodes") const;
+  void dumpNodes(const std::string& name = "nodes") const;
+
+ private:
+  odb::dbBlock* getBlock() const;
+  odb::dbTech* getTech() const;
+
+  void addStrapNodes(odb::dbSBox* box);
+  void addViaNodes(odb::dbSBox* box);
+
+  void sortShapes();
+
+  void generateRoutingLayerShapesAndNodes();
+  void generateCutLayerNodes();
+  void generateTopLayerFillerNodes();
+  void sortNodes();
+  void cleanupNodes();
+  void cleanupOverlappingNodes(NodePtrMap<Connection>& connection_map);
+  void mergeNodes(NodePtrMap<Connection>& connection_map);
+
+  void connectLayerNodes();
+  void cleanupConnections();
+  void cleanupInvalidConnections();
+  void cleanupDuplicateConnections();
+  void sortConnections();
+
+  void removeNodes(std::set<Node*>& removes,
+                   odb::dbTechLayer* layer,
+                   std::vector<std::unique_ptr<Node>>& nodes,
+                   const NodePtrMap<Connection>& connection_map);
+  void removeConnections(std::set<Connection*>& removes);
+
+  int getEffectiveNumberOfCuts(const odb::dbShape& shape) const;
+
+  void copy(Node* keep, Node* remove, NodePtrMap<Connection>& connection_map);
+
+  std::set<Node*> getSharedShapeNodes() const;
+
+  Polygon90 rectToPolygon(const odb::Rect& rect) const;
+  LayerMap<Polygon90Set> generatePolygonsFromSWire(odb::dbSWire* wire);
+  LayerMap<Polygon90Set> generatePolygonsFromITerms(
+      std::vector<TerminalNode*>& terminals);
+  LayerMap<Polygon90Set> generatePolygonsFromBTerms(
+      std::vector<TerminalNode*>& terminals);
+  void processPolygonToRectangles(
+      odb::dbTechLayer* layer,
+      const Polygon90& polygon,
+      const TerminalTree& terminals,
+      std::vector<std::unique_ptr<Shape>>& new_shapes,
+      std::vector<std::unique_ptr<Node>>& new_nodes,
+      std::map<Shape*, std::set<Node*>>& terminal_connections);
+  void generateCutNodesForSBox(
+      odb::dbSBox* box,
+      bool single_via,
+      std::vector<std::unique_ptr<Node>>& new_nodes,
+      std::vector<std::unique_ptr<Connection>>& new_connections);
+  TerminalTree getTerminalTree(
+      const std::vector<TerminalNode*>& terminals) const;
+  ShapeTree getShapeTree(odb::dbTechLayer* layer) const;
+  NodeTree getNodeTree(odb::dbTechLayer* layer) const;
+
+  void initMinimumNodePitch();
+
+  void recoverMemory();
+
+  void reset();
+
+  odb::dbNet* net_;
+
+  utl::Logger* logger_;
+
+  bool floorplanning_;
+
+  LayerMap<std::vector<std::unique_ptr<Shape>>> shapes_;
+  LayerMap<std::vector<std::unique_ptr<Node>>> nodes_;
+
+  std::vector<std::unique_ptr<Connection>> connections_;
+
+  std::vector<std::unique_ptr<ITermNode>> iterm_nodes_;
+  std::vector<std::unique_ptr<BPinNode>> bpin_nodes_;
+
+  std::map<odb::dbTechLayer*, int> min_node_pitch_;
+
+  static constexpr int min_node_pitch_multiplier_ = 10;
+  static constexpr double min_node_pitch_um_ = 10.0;
```

Comment:
"magic numbers", the pitch multiplier is "kind of like" the row multiplier in the previous implementation (defaulted to 5), 10 tracks is much less than 5 row heights, so this seemed reasonable, but I'm not wed to these numbers.
10um is just to ensure the top layer has enough nodes to intersect with the voltage sources (10 happen to the smallest in the test cases), but it's possible that is not needed any longer since the voltages sources were updated to "snap" to the nearest node.